File: bubbly/builtin/release.go
package builtin

type ReleaseStatus string

const (
	ReadyReleaseStatus   ReleaseStatus = "READY"
	PendingReleaseStatus ReleaseStatus = "PENDING"
	BlockedReleaseStatus ReleaseStatus = "BLOCKED"
)

func ReleaseStatusByStages(release Release) ReleaseStatus {
	var status ReleaseStatus = ReadyReleaseStatus
	for _, stage := range release.ReleaseStage {
		stageStatus := ReleaseStageStatus(stage)
		switch stageStatus {
		case BlockedReleaseStatus:
			return BlockedReleaseStatus
		case PendingReleaseStatus:
			status = PendingReleaseStatus
		}
	}
	return status
}

func ReleaseStageStatus(stage ReleaseStage) ReleaseStatus {
	var status ReleaseStatus = ReadyReleaseStatus
	for _, criteria := range stage.ReleaseCriteria {
		criteriaStatus := ReleaseCriteriaStatus(criteria)
		switch criteriaStatus {
		case BlockedReleaseStatus:
			return BlockedReleaseStatus
		case PendingReleaseStatus:
			status = PendingReleaseStatus
		}
	}
	return status
}

func ReleaseCriteriaStatus(criteria ReleaseCriteria) ReleaseStatus {
	if len(criteria.ReleaseEntry) == 0 {
		return PendingReleaseStatus
	}
	// We only care about the first (or latest) release entry
	if !criteria.ReleaseEntry[0].Result {
		return BlockedReleaseStatus
	}
	return ReadyReleaseStatus
}

File: ui/src/utils/release.ts

import type { Release } from '$schema/schema_gen';
import { ReleaseStatus } from '$schema/schema_gen';

export const releaseStatusColor = (release: Release): string => {
	switch (release.status) {
		case ReleaseStatus.ready:
			return 'green';
		case ReleaseStatus.blocked:
			return 'red';
		case ReleaseStatus.pending:
			return 'yellow';
		default:
			return 'gray';
	}
};

export const releaseStatusText = (release: Release): string => {
	switch (release.status) {
		case ReleaseStatus.ready:
			return 'Ready';
		case ReleaseStatus.blocked:
			return 'Blocked';
		case ReleaseStatus.pending:
			return 'Pending';
		default:
			return 'Status Unknown';
	}
};
